File: php/tests/Unit/KeyGeneratorTest.php
<?php

declare(strict_types=1);

namespace RespBench\Tests\Unit;

use PHPUnit\Framework\TestCase;
use RespBench\Config\KeyspaceConfig;
use RespBench\Engine\KeyGenerator;

final class KeyGeneratorTest extends TestCase
{
    public function testGeneratesSequentialKeys(): void
    {
        $config = new KeyspaceConfig(keysCount: 100, keyPrefix: 'test:');
        $gen = new KeyGenerator($config);

        $key1 = $gen->nextKey();
        $key2 = $gen->nextKey();

        self::assertStringStartsWith('test:', $key1);
        self::assertStringStartsWith('test:', $key2);
        self::assertNotSame($key1, $key2);
    }

    public function testSequentialKeysWrapAround(): void
    {
        $config = new KeyspaceConfig(keysCount: 3, keyPrefix: 'test:');
        $gen = new KeyGenerator($config);

        $keys = [];
        for ($i = 0; $i < 6; $i++) {
            $keys[] = $gen->nextKey();
        }

        // Keys should wrap: 0, 1, 2, 0, 1, 2
        self::assertSame($keys[0], $keys[3]);
        self::assertSame($keys[1], $keys[4]);
        self::assertSame($keys[2], $keys[5]);
    }

    public function testGeneratesUniformRandomKeys(): void
    {
        $config = new KeyspaceConfig(
            keysCount: 1000,
            keyPrefix: 'rand:',
            generationAlg: 'uniform_rand',
            seed: 12345,
        );
        $gen = new KeyGenerator($config);

        $keys = [];
        for ($i = 0; $i < 100; $i++) {
            $keys[] = $gen->nextKey();
        }

        $unique = array_unique($keys);
        self::assertGreaterThan(50, count($unique));
    }

    public function testResetRestartsSequentialCounter(): void
    {
        $config = new KeyspaceConfig(keysCount: 100, keyPrefix: 'test:');
        $gen = new KeyGenerator($config);

        $first1 = $gen->nextKey();
        $gen->nextKey();
        $gen->reset();
        $first2 = $gen->nextKey();

        self::assertSame($first1, $first2);
    }

    public function testResetRestartsRandomSequence(): void
    {
        $config = new KeyspaceConfig(
            keysCount: 1000,
            keyPrefix: 'rand:',
            generationAlg: 'uniform_rand',
            seed: 12345,
        );
        $gen = new KeyGenerator($config);

        $first = [];
        for ($i = 0; $i < 10; $i++) {
            $first[] = $gen->nextKey();
        }
        $gen->reset();
        $second = [];
        for ($i = 0; $i < 10; $i++) {
            $second[] = $gen->nextKey();
        }

        self::assertSame($first, $second);
    }

    public function testKeyFormatZeroPadsToKeySize(): void
    {
        $config = new KeyspaceConfig(
            keysCount: 100,
            keySizeBytes: 16,
            keyPrefix: 'bench:',
        );
        $gen = new KeyGenerator($config);

        $key = $gen->nextKey();

        // prefix "bench:" (6) + padding width max(16-6,1)=10 -> "bench:0000000000"
        self::assertSame('bench:0000000000', $key);
        self::assertSame(16, strlen($key));
    }

    public function testUniformRandMatchesJavaRandomSequence(): void
    {
        // With uniform_rand + seed, keys are prefix + JavaRandom.nextInt(keysCount).
        $config = new KeyspaceConfig(
            keysCount: 1000,
            keySizeBytes: 16,
            keyPrefix: 'bench:',
            generationAlg: 'uniform_rand',
            seed: 0,
        );
        $gen = new KeyGenerator($config);

        $first = $gen->nextKey();
        // JavaRandom(0).nextInt(1000) == 360 (verified against Java canonical LCG).
        self::assertSame('bench:0000000360', $first);
    }

    public function testSequentialWorkersPartitionKeyspaceWithoutDuplication(): void
    {
        // Regression: previously every forked worker restarted at 0, so N workers
        // wrote only keys_count/N distinct keys. Now worker i strides i, i+N, ...
        // and the union covers the full keyspace exactly once.
        $config = new KeyspaceConfig(
            keysCount: 20,
            keySizeBytes: 8,
            keyPrefix: 'k:',
            generationAlg: 'sequential_int',
        );

        $workerCount = 4;
        $perWorker = 5; // 4 * 5 = 20 = keys_count
        $all = [];
        for ($w = 0; $w < $workerCount; $w++) {
            $gen = KeyGenerator::forWorker($config, $w, $workerCount);
            for ($i = 0; $i < $perWorker; $i++) {
                $all[] = $gen->nextKey();
            }
        }

        self::assertCount(20, $all);
        self::assertCount(20, array_unique($all), 'workers must not duplicate keys');
    }

    public function testSingleWorkerSequentialWalksFullKeyspace(): void
    {
        $config = new KeyspaceConfig(
            keysCount: 5,
            keySizeBytes: 8,
            keyPrefix: 'k:',
            generationAlg: 'sequential_int',
        );
        $gen = KeyGenerator::forWorker($config, 0, 1);

        $keys = [];
        for ($i = 0; $i < 5; $i++) {
            $keys[] = $gen->nextKey();
        }
        self::assertCount(5, array_unique($keys));
    }
}

File: go/internal/engine/keygenerator_test.go
package engine

import (
	"testing"

	"github.com/resp-bench/go/internal/config"
)

func seq(count, size int, prefix string) config.KeyspaceConfig {
	return config.KeyspaceConfig{
		KeysCount:     count,
		KeySizeBytes:  size,
		KeyPrefix:     prefix,
		GenerationAlg: "sequential_int",
	}
}

// TestSequentialWrapsAround: a single worker emits 0..N-1 then wraps.
func TestSequentialWrapsAround(t *testing.T) {
	// prefix "test:" (5) + key_size 16 → pad width = 16-5 = 11 → "%011d".
	set := NewKeyGenSet(seq(3, 16, "test:"))
	kg := set.ForWorker(0)
	want := []string{"test:00000000000", "test:00000000001", "test:00000000002", "test:00000000000"}
	for i, w := range want {
		if got := kg.NextKey(); got != w {
			t.Fatalf("key %d: got %q, want %q", i, got, w)
		}
	}
}

// TestSequentialPartitionAcrossWorkers: N workers share one counter, so their
// combined output covers the keyspace exactly once with no duplicates.
func TestSequentialPartitionAcrossWorkers(t *testing.T) {
	const workers = 4
	const keys = 5
	set := NewKeyGenSet(seq(keys*workers, 16, "k:"))

	gens := make([]*KeyGenerator, workers)
	for i := range gens {
		gens[i] = set.ForWorker(i)
	}

	seen := map[string]bool{}
	// Each worker pulls `keys` keys; total = workers*keys distinct keys.
	for i := 0; i < keys; i++ {
		for w := 0; w < workers; w++ {
			k := gens[w].NextKey()
			if seen[k] {
				t.Fatalf("duplicate key %q across workers", k)
			}
			seen[k] = true
		}
	}
	if len(seen) != workers*keys {
		t.Fatalf("expected %d distinct keys, got %d", workers*keys, len(seen))
	}
}

// TestUniformRandReproducible: same seed + worker index → same key stream.
func TestUniformRandReproducible(t *testing.T) {
	seed := int64(12345)
	cfg := config.KeyspaceConfig{
		KeysCount:     1000,
		KeySizeBytes:  16,
		KeyPrefix:     "r:",
		GenerationAlg: "uniform_rand",
		Seed:          &seed,
	}
	a := NewKeyGenSet(cfg).ForWorker(0)
	b := NewKeyGenSet(cfg).ForWorker(0)
	for i := 0; i < 100; i++ {
		if x, y := a.NextKey(), b.NextKey(); x != y {
			t.Fatalf("key %d diverged: %q vs %q", i, x, y)
		}
	}
}

// TestUniformRandSeedZeroAnchor: with seed 0 and 1000 keys, the first index is
// JavaRandom(0).nextInt(1000) == 360 (verified against the canonical Java LCG).
func TestUniformRandSeedZeroAnchor(t *testing.T) {
	seed := int64(0)
	cfg := config.KeyspaceConfig{
		KeysCount:     1000,
		KeySizeBytes:  8,
		KeyPrefix:     "k:",
		GenerationAlg: "uniform_rand",
		Seed:          &seed,
	}
	kg := NewKeyGenSet(cfg).ForWorker(0)
	// prefix "k:" (2) + key_size 8 → pad width = 6 → "%06d"; index 360 → "000360".
	if got := kg.NextKey(); got != "k:000360" {
		t.Fatalf("first uniform_rand key: got %q, want %q", got, "k:000360")
	}
}
